coverage/hare_coverage.py

```diff
@@ -0,0 +1,319 @@
+                    filename='coverage.log')
+
+
+class InvalidDirectoryPath(BaseException):
+
+    """Used to generate the invalid path exception."""
+
+    pass
+
+
+class FailedToGenerateReport(BaseException):
+
+    """Used to raise exception while report generation failed."""
+
+    pass
+
+
+class DirectoryScanner:
+
+    """
+    Scans the directory according to the given directory name. It never return files that are rejected by the
+    filter with 'omit' keyword argument.
+    """
+
+    def __init__(self, pattern: str, callback: Optional[Callable[[str, str], bool]] = None, **kwargs):
+        """
+        Initialization.
+        :param pattern: Directory name to scan
+        :param callback: callback for comparison
+        :param kwargs: Keyword arguments for the filtering.
+        """
+        self.pattern = pattern
+        self.callback = callback
+        if self.callback is None:
+            self.callback = self.comparator
+        self.omit_kwargs = kwargs.get("omit", "py3venv build").split()
+
+    def comparator(self, dir_name: str) -> bool:
+        """
+        Comparator.
+        :param dir_name: directory name
+        :return: boolean
+        """
+        return dir_name == self.pattern
+
+    def omit_dir(self, dir_name: str) -> bool:
+        """
+        Filter out the omitted directories.
+        :param dir_name: directory name
+        :return: boolean
+        """
+        for word in self.omit_kwargs:
+            if dir_name.find(word) != -1:
+                return True
+        return False
+
+    def search_dir(self, dir_path: str) -> List[str]:
+        """
+        Strategy for scanning directories.
+        :param dir_path: directory path
+        :return: List of the file match to given pattern
+        """
+        result_list: List[str] = []
+        for sub_dir, dirs, _ in walk(dir_path):
+            for directory in dirs:
+                directory_path: str = sub_dir + sep + directory
+                if self.omit_dir(directory_path):
+                    continue
+                result_list.extend([ele for ele in self.search_dir(directory_path) if ele not in result_list])
+                if self.callback(directory):
+                    if directory_path not in result_list:
+                        result_list.append(directory_path)
+        return result_list
+
+
+class Utility:
+
+    """Here is the utility class which has static utility helper methods."""
+
+    def run_cmd(self, command: str, delimiter: Optional[str] = None, max_split: int = -1) -> None:
+        """
+        Execute command and return the result.
+        :param command: Command in string format
+        :param delimiter: delimiter (default is space)
+        :param max_split: maximum number split
+        :return: tuple(stdout, stderr)
+        """
+        logging.debug(f"Executing Command : {command!a}")
+        process: Popen[str] = Popen(command.split(sep=delimiter, maxsplit=max_split),

```

Comment:
mypy says:
Incompatible types in assignment (expression has type "Popen[bytes]", variable has type "Popen[str]")

Reply:
Resolved